linkerd/protocol/h2/src/main/scala/io/buoyant/linkerd/protocol/h2/H2RequestActiveTracer.scala

```diff
@@ -0,0 +1,147 @@
+  private[this] val AddRouterContextHeader = "l5d-add-context"
+
+  private[this] def getRequestTraceResponse(resp: Response, stopwatch: Stopwatch.Elapsed):Future[Response] = {
+    val routerCtxF = RouterContextBuilder(
+      routerLabel,
+      stopwatch,
+      DstPathCtx.current,
+      DstBoundCtx.current,
+      endpoint,
+      namers,
+      dtab)
+
+    routerCtxF.map { ctx =>
+      // format current router context
+      val content = s"""|
+                        |${ctx.formatRouterContext.trim}
+                        |""".stripMargin
+      if (resp.stream.isEmpty) {
+        // add router context data frame to empty stream
+        Response(resp.headers, Stream.const(Frame.Data(content, eos = true)))
+      } else {
+        val newStream = resp.stream.flatMap {
+          // read every frame of a stream
+          case trailerFrame: Trailers =>
+            // on trailers frame, prepend a data frame containing router context to trailer
+            val routerCtxFrame = Frame.Data(content, eos = false)
+            Seq(routerCtxFrame, trailerFrame)
+          case dataFrame: Data if dataFrame.isEnd =>
+            // on data frame with EOS flag to true, prepend new data frame with router context to
+            // EOS data frame
+            val newDataFrame = Frame.Data.copy(dataFrame, eos = false)
+            val routerCtxFrame = Data(content, eos = true)
+            Seq(newDataFrame, routerCtxFrame, Trailers())
+          case frame =>
+            // do nothing if we find a frame that doesn't indicate the end of a stream
+            Seq(frame)
+        }
+        Response(resp.headers, newStream)
+      }
+    }
+
+  }
+
+  override def apply(
+    request: Request,
+    service: Service[Request, Response]
+  ): Future[Response] = {
+    val maxForwards = Try(request.headers.get(MaxForwards).map(_.toInt))
+    val isAddRouterCtx = request.headers.get(AddRouterContextHeader) match {
+      case Some(v) => StringUtil.toBoolean(v)
+      case None => false
+    }
+
+    (request.method, maxForwards, isAddRouterCtx) match {
+      case (Method.Trace, Throw(_: NumberFormatException), _) =>
+        // MaxForwards is unparsable send a bad request response
+        Future.value(Response(Status.BadRequest, Stream.empty))
+      case (Method.Trace, Return(Some(0)), true) =>
+        // MaxForwards header is 0 and the l5d-add-context header is present
+        // returns a response with router context
+        val stopwatch = Stopwatch.start()
+        getRequestTraceResponse(Response(request.headers, Stream.empty), stopwatch)
+      case (Method.Trace, Return(Some(0)), false) =>
+        // returns a response with no router context
+        Future.value(Response(Status.Ok, Stream.empty))
+      case (Method.Trace, Return(Some(num)), true) if num > 0 =>
+        // Decrement Max-Forwards header. forward H2 request downstream
+        val stopwatch = Stopwatch.start()
+        request.headers.set(MaxForwards, (num - 1).toString)
```

Comment:
The H2 request type is better about mutability than H1 and so we can do something nicer here:

```
val req1 = request.dup()
req1.headers.set(...)
service(req1)
```
since we never mutate the original request, we never have to worry about un-mutating it!